hushline/routes.py

```diff
@@ -154,6 +175,15 @@ def submit_message(username: str) -> Response | str:
 
             captcha_answer = request.form.get("captcha_answer", "")
             if not validate_captcha(captcha_answer):
+                # Encrypt the message and store it in the session
+                scope = "submit_message"
+                salt = generate_salt()
```

Comment:
I like that part of the secret here lives only in the session, but do we still need a scope if we have a unique salt? My intuition says that the salt binds the scope tighter than the "submit_message" scope.

Reply:
@jeremywmoore In a situation where the salt is guaranteed to only be used in one place, that may be enough. But, it can get complicated to make that determination. For instance, it's being injected into the session, so access to it is given to other parts of the system (and potentially users). As a result, we can 1. carefully make sure nothing uses it in any other situation; 2. ensure that it can't be harmful if it can somehow be injected into other contexts; or, 3. encode its purpose as a scope. The explicit scope is a much easier, more localized, and less fragile design choice, since it removes the cognitive burden of all that tracking, double-checking, and repeat security analyses when new features are added.

Here are some related resources on the topic:
https://www.cs.cornell.edu/andru/papers/iflow-tosem.pdf
https://www.rfc-editor.org/rfc/rfc1457

Reply:
Thanks for the explanation @rmlibre! This feels like sufficient reason to keep the scope in addition to the salt.